src/qb4-packw/gen/qb4-packw-x16c8-gemm-goi-scalar.c

```diff
@@ -0,0 +1,381 @@
+// clang-format off
+// Auto-generated file. Do not edit!
+//   Template: src/qb4-packw/kr-scalar.c.in
+//   Generator: tools/xngen
+//
+// Copyright 2023 Google LLC
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include <assert.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "src/xnnpack/packw.h"
+
+static XNN_INTRINSIC void xnn_packed2planar(
+  int32_t* ksum,
+  const uint8_t* weights,
+  int8_t* out,
+  int32_t vkernel_zero_point)
+{
+  const uint32_t s_v0 = (((const uint32_t*)weights)[0] ^ vkernel_zero_point);
+  const int8_t v0 = (((const int8_t*)&s_v0))[0] << 4;
+  const int8_t v1 = (((const int8_t*)&s_v0))[0] & 0xF0;
+  const int8_t v2 = (((const int8_t*)&s_v0))[1] << 4;
+  const int8_t v3 = (((const int8_t*)&s_v0))[1] & 0xF0;
+  const int8_t v4 = (((const int8_t*)&s_v0))[2] << 4;
+  const int8_t v5 = (((const int8_t*)&s_v0))[2] & 0xF0;
+  const int8_t v6 = (((const int8_t*)&s_v0))[3] << 4;
+  const int8_t v7 = (((const int8_t*)&s_v0))[3] & 0xF0;
+  const uint32_t s_v1 = (((const uint32_t*)weights)[1] ^ vkernel_zero_point);
+  const int8_t v8 = (((const int8_t*)&s_v1))[0] << 4;
+  const int8_t v9 = (((const int8_t*)&s_v1))[0] & 0xF0;
+  const int8_t v10 = (((const int8_t*)&s_v1))[1] << 4;
+  const int8_t v11 = (((const int8_t*)&s_v1))[1] & 0xF0;
+  const int8_t v12 = (((const int8_t*)&s_v1))[2] << 4;
+  const int8_t v13 = (((const int8_t*)&s_v1))[2] & 0xF0;
+  const int8_t v14 = (((const int8_t*)&s_v1))[3] << 4;
+  const int8_t v15 = (((const int8_t*)&s_v1))[3] & 0xF0;
+
+  (*ksum) += (int32_t) (v0);
+  (*ksum) += (int32_t) (v1);
+  (*ksum) += (int32_t) (v2);
+  (*ksum) += (int32_t) (v3);
+  (*ksum) += (int32_t) (v4);
+  (*ksum) += (int32_t) (v5);
+  (*ksum) += (int32_t) (v6);
+  (*ksum) += (int32_t) (v7);
+  (*ksum) += (int32_t) (v8);
+  (*ksum) += (int32_t) (v9);
+  (*ksum) += (int32_t) (v10);
+  (*ksum) += (int32_t) (v11);
+  (*ksum) += (int32_t) (v12);
+  (*ksum) += (int32_t) (v13);
+  (*ksum) += (int32_t) (v14);
+  (*ksum) += (int32_t) (v15);
+
+  out[0] = (((uint8_t) v0) >> 4 | (v8));
+  out[1] = (((uint8_t) v1) >> 4 | (v9));
+  out[2] = (((uint8_t) v2) >> 4 | (v10));
+  out[3] = (((uint8_t) v3) >> 4 | (v11));
+  out[4] = (((uint8_t) v4) >> 4 | (v12));
+  out[5] = (((uint8_t) v5) >> 4 | (v13));
+  out[6] = (((uint8_t) v6) >> 4 | (v14));
+  out[7] = (((uint8_t) v7) >> 4 | (v15));
+}
+
+void xnn_qb4_packw_gemm_goi_ukernel_x16c8__scalar(
+  size_t g,
+  size_t nc,
+  size_t kc,
+  size_t nr,
+  size_t kr,
+  size_t sr,
+  size_t bl,
+  const uint8_t* weights,
+  const int32_t* bias,
+  const void* scale,
+  int8_t* packed_weights,
+  size_t extra_bytes_bl,
+  size_t extra_bytes_n,
+  const void* params)
```

Comment:
This kernel either needs to be attributed `XNN_OOB_READS`, or it has a bug that causes it to read out of bounds:
```
[ RUN      ] qb4_packw/XnnTestQB4.nc_gt_nr/xnn_qb4_packw_gemm_goi_ukernel_x16c8__scalar
=================================================================
==10==ERROR: AddressSanitizer: use-after-poison on address 0x59b00154 at pc 0x006d9f94 bp 0x407ff040 sp 0x407ff03c
READ of size 4 at 0x59b00154 thread T0
```
(call stack is useless unfortunately)

Reply:
i'll figure this one out thanks!

Reply:
the issue was an out of bounds read of the bias, since our packing kernels also do a one shot bias fill for packing. I marked the packing kernels as XNN_OOB_READ.